### PR TITLE
10.1.Create Account with Country and City fields

### PR DESCRIPTION
**Pull request by xalivan**

### CODE REVIEW

File: src/main/java/com/speaker/convertors/AccountConverter.java [hunk 1]
```diff
@@ -48,6 +49,14 @@ public Account convertToAccount(AccountDTO accountDTO) {
...
                 .lastName(accountDTO.getLastName())
                 .age(accountDTO.getAge())
+                .country(convertToCountry(accountDTO.getCountry()))
+                .build();
+    }
+    public Country convertToCountry(CountryDTO countryDTO) {
+        return Country.builder()
+                .id(countryDTO.getId())
+                .name(countryDTO.getName())
+                .city(countryDTO.getCity())
```
Comment (otkalenko) [inline src/main/java/com/speaker/convertors/AccountConverter.java:59]: must be CityDTO

File: src/main/java/com/speaker/service/AccountServiceImpl.java [hunk 2]
```diff
@@ -29,7 +32,36 @@
...
+    public boolean create(AccountDTO accountDTO) {
+        Map<CityName, Pair<Integer, Integer>> cityNamePairMap = mapCountryAndCity.get(accountDTO.getCountry().getName());
+        Pair<Integer, Integer> integerIntegerPair = cityNamePairMap.get(accountDTO.getCountry().getCityDTO().getName());
+        return accountRepository.insert(accountConverter
+                .convertToAccount(accountDTO, integerIntegerPair.getFirst(), integerIntegerPair.getSecond())) > 0;
+    }
+
+    @PostConstruct
+    private void getAllCountryAndCity() {
+        List<Country> allCountryAndCity = accountRepository.findAllCountryAndCity();
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/AccountServiceImpl.java:44]: not need create new variable `List<Country> allCountryAndCity` can just:
```java
mapCountryAndCity= accountRepository.findAllCountryAndCity().stream()
                .map(country -> new Pair<CountryName, Pair<CityName, Pair<Integer, City>>>(country.getName(),
                        new Pair<CityName, Pair<Integer, City>>(country.getCity().getName(),
                                new Pair<Integer, City>(country.getId(), country.getCity()))
                )).collect(Collectors.groupingBy(Pair::getFirst, Collectors.mapping(Pair::getSecond,
                        Collectors.toMap(Pair::getFirst, this::convertToIds, (city1, city2) -> city1))));
```

File: src/main/java/com/speaker/service/AccountServiceImpl.java [hunk 3]
```diff
@@ -2,17 +2,20 @@
...
 import org.springframework.stereotype.Service;
 
+import javax.annotation.PostConstruct;
 import java.util.List;
+import java.util.Map;
 import java.util.stream.Collectors;
 
 @Service
-@RequiredArgsConstructor
+@Data
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/AccountServiceImpl.java:16]: data? why?

Comment (otkalenko) [reply src/main/java/com/speaker/service/AccountServiceImpl.java:19]: ?

Comment (otkalenko) [reply src/main/java/com/speaker/service/AccountServiceImpl.java:23]: ?

File: src/test/java/com/speaker/service/AccountServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    @BeforeEach
+    public void prepare() {
+        when(accountRepository.findAllCountryAndCity())
+                .thenReturn(List.of(generateAccount("present",
+                        CountryName.UKRAINE, "present", CityName.KYIV).getCountry()));
+        this.accountService.init();
+    }
+
+    @Test
+    void findAll() {
```
Comment (otkalenko) [inline src/test/java/com/speaker/service/AccountServiceImplTest.java:45]: please add access modification

File: src/test/java/com/speaker/service/AccountServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Test
+    void findAllFriendsByAccountId() {
+        Account account = generateAccount("present", CountryName.UKRAINE, "present", CityName.KYIV);
+        when(accountRepository.findAllFriendsByAccountId(1)).thenReturn(List.of(account));
+        assertThat(accountService.findAllFriendsByAccountId(1), is(List.of(account)));
+    }
+
+    @Test
+    void createSuccess() {
```
Comment (otkalenko) [inline src/test/java/com/speaker/service/AccountServiceImplTest.java:61]: accountCreatedSuccess

File: src/test/java/com/speaker/service/AccountServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        Account account = generateAccount("present", CountryName.UKRAINE, "present", CityName.KYIV);
+        AccountDTO accountDTO = generateAccountDTO("present", CountryName.UKRAINE, "present", CityName.KYIV);
+        when(accountConverter.convertToAccount(accountDTO, accountDTO.getCountry().getId(),
+                accountDTO.getCountry().getCityDTO().getId())).thenReturn(account);
+        when(accountRepository.insert(account)).thenReturn(1);
+        assertThat(accountService.create(accountDTO), is(Response.TRUE));
+    }
+
+    @Test
+    void createIfCountryNull() {
```
Comment (otkalenko) [inline src/test/java/com/speaker/service/AccountServiceImplTest.java:71]: accountNotCreatedWhenCityIsNull

File: src/test/java/com/speaker/utils/AccountDTOGenerator.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import com.speaker.dto.AccountDTO;
+import com.speaker.dto.CityDTO;
+import com.speaker.dto.CountryDTO;
+import com.speaker.entities.CityName;
+import com.speaker.entities.CountryName;
+import lombok.RequiredArgsConstructor;
+
+import java.util.Random;
+
+@RequiredArgsConstructor
```
Comment (otkalenko) [inline src/test/java/com/speaker/utils/AccountDTOGenerator.java:12]: Is it a real need?

File: src/test/java/com/speaker/utils/AccountGenerator.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
+package com.speaker.utils;
+
+import com.speaker.entities.*;
+import lombok.RequiredArgsConstructor;
+
+import java.util.Random;
+
+@RequiredArgsConstructor
```
Comment (otkalenko) [inline src/test/java/com/speaker/utils/AccountGenerator.java:8]: the same

File: src/test/java/com/speaker/service/AccountServiceImplTest.java (hunk 4 above)
Comment (otkalenko) [reply src/test/java/com/speaker/service/AccountServiceImplTest.java:45]: find all accounts?

File: src/main/java/com/speaker/service/AccountServiceImpl.java [hunk 9]
```diff
@@ -29,7 +49,33 @@ public List<Account> findAllFriendsByAccountId(int accountId) {
...
+        }
+        return Response.FALSE;
+    }
+
+    private Pair<Integer, Integer> convertToIds(Pair<CityName, Pair<Integer, City>> entry) {
+        return new Pair<>(entry.getSecond().getFirst(), entry.getSecond().getSecond().getId());
+    }
+
+    @Data
+    class Pair<A, B> {
```
Comment (otkalenko) [inline src/main/java/com/speaker/service/AccountServiceImpl.java:72]: please extract in separately entity, maybe we will use in the next cases

File: src/main/java/com/speaker/controller/AccountController.java [hunk 10]
```diff
@@ -20,8 +21,8 @@ public class AccountController {
 
 
     @PostMapping
-    public ResponseEntity<Integer> create(@RequestBody AccountDTO accountDTO) {
-        return ResponseEntity.ok(accountService.create(accountDTO));
+    public Response create(@RequestBody AccountDTO accountDTO) {
+        return accountService.create(accountDTO);
```
Comment (otkalenko) [inline src/main/java/com/speaker/controller/AccountController.java:25]: ResponseEntity.ok(accountService.create(accountDTO))

File: src/test/java/com/speaker/utils/AccountDTOGenerator.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+import com.speaker.entities.CityName;
+import com.speaker.entities.CountryName;
+
+import java.util.Random;
+
+public class AccountDTOGenerator {
+    private final static Random RANDOM = new Random();
+
+    public static AccountDTO generateAccountDTO(CountryDTO country) {
+        if (country == null) {
```
Comment (otkalenko) [inline src/test/java/com/speaker/utils/AccountDTOGenerator.java:15]: looks like bulky, let's remove duplicate code

```java
return AccountDTO.builder()
                    .id(RANDOM.nextInt(10))
                    .name("Ivan")
                    .lastName("Mazur")
                    .age(RANDOM.nextInt(10))
                    .country(Optional.ofNullable(country)
                            .map(countryDTO-> generateCountryDTO(countryDTO.getName(), countryDTO.getCityDTO()))
                            .orElse(null))
                    .build();
```

File: src/test/java/com/speaker/utils/AccountDTOGenerator.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+                    .name("Ivan")
+                    .lastName("Mazur")
+                    .age(RANDOM.nextInt(10))
+                    .country(generateCountryDTO(country.getName(), country.getCityDTO()))
+                    .build();
+        }
+    }
+
+    public static CountryDTO generateCountryDTO(CountryName countryName, CityDTO city) {
+        if (city == null) {
```
Comment (otkalenko) [inline src/test/java/com/speaker/utils/AccountDTOGenerator.java:35]: the same

File: src/test/java/com/speaker/service/AccountServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        assertThat(accountService.create(accountDTO), is(Response.FALSE));
+    }
+
+    @Test
+    public void accountNotCreatedWhenCityIsNull() {
+        AccountDTO accountDTO = generateAccountDTO(generateCountryDTO(CountryName.UKRAINE, null));
+        assertThat(accountService.create(accountDTO), is(Response.FALSE));
+    }
+
+    @Test()
```
Comment (otkalenko) [inline src/test/java/com/speaker/service/AccountServiceImplTest.java:87]: do you need parentheses here?

File: src/test/java/com/speaker/service/AccountServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+        this.accountService.init();
+    }
+
+    @Test
+    public void findAllAccounts() {
+        Account account = generateAccount(generateCountry(CountryName.UKRAINE, generateCity(CityName.KYIV)));
+        when(accountConverter.convertToAccountDTO(account))
+                .thenReturn(generateAccountDTO(generateCountryDTO(CountryName.UKRAINE, generateCityDTO(CityName.KYIV))));
+        when(accountRepository.findAll()).thenReturn(List.of(account));
+        assertThat(accountService.findAll(), containsInAnyOrder(accountConverter.convertToAccountDTO(account)));
```
Comment (otkalenko) [inline src/test/java/com/speaker/service/AccountServiceImplTest.java:49]: You must expect objects that you generate. If mock `accountConverter.convertToAccountDTO(` will return nothing, you will expect nothing, it's a wrong test.

File: src/main/java/com/speaker/convertors/AccountConverter.java [hunk 15]
```diff
@@ -43,11 +45,29 @@ public AccountDTO convertToAccountDTO(Account account) {
...
+    public Account convertToAccount(AccountDTO accountDTO, int countryId, int cityId) {
         return Account.builder()
                 .name(accountDTO.getName())
                 .lastName(accountDTO.getLastName())
                 .age(accountDTO.getAge())
+                .country(convertToCountry(accountDTO.getCountry(), countryId, cityId))
                 .build();
     }
+
+    public Country convertToCountry(CountryDTO countryDTO, int countryId, int cityId) {
```
[the remark this answers is not in the document]
Comment (xalivan) [reply src/main/java/com/speaker/convertors/AccountConverter.java]: Comments accepted, changes made. A new CountryConverter.class has been created

**Review by otkalenko — APPROVED** [review]

File: src/main/java/com/speaker/convertors/AccountConverter.java (hunk 1 above)
Comment (xalivan) [reply src/main/java/com/speaker/convertors/AccountConverter.java]: Test

**Pull request merged**